src/Microsoft.OData.Core/JsonLight/ODataJsonLightPropertyAndValueDeserializer.cs

```diff
@@ -1468,15 +1470,192 @@ private object ReadEnumValue(bool insideJsonObjectValue, IEdmEnumTypeReference e
+        /// <param name="payloadTypeName">The type name read from the payload.</param>
+        /// <param name="serializationTypeNameAnnotation">The serialization type name for the collection value (possibly null).</param>
+        /// <param name="duplicatePropertyNamesChecker">The duplicate property names checker to use - this is always initialized as necessary, do not clear.</param>
+        /// <returns>The value of the resource value.</returns>
+        /// <remarks>
+        /// Pre-Condition:  JsonNodeType.Property - the first property of the resource value object, or the second one if the first one was odata.type.
+        ///                 JsonNodeType.EndObject - the end object of the resource value object.
+        /// Post-Condition: almost anything - the node after the resource value (after the EndObject)
+        /// </remarks>
+        private ODataResourceValue ReadResourceValue(
+            bool insideJsonObjectValue,
+            bool insideResourceValue,
+            string propertyName,
+            IEdmStructuredTypeReference structuredTypeReference,
+            string payloadTypeName,
+            PropertyAndAnnotationCollector propertyAndAnnotationCollector)
+        {
+            if (!insideJsonObjectValue && !insideResourceValue)
+            {
+                if (this.JsonReader.NodeType != JsonNodeType.StartObject)
+                {
+                    string typeName = structuredTypeReference != null ? structuredTypeReference.FullName() : payloadTypeName;
+                    throw new ODataException(
+                        string.Format(CultureInfo.InvariantCulture,
+                        "The property with name '{0}' was found with a value node of type '{1}'; however, a resource value of type '{2}' was expected.",
+                        propertyName, this.JsonReader.NodeType, typeName));
+                }
+
+                this.JsonReader.Read();
+            }
+
+            return this.ReadResourceValue(structuredTypeReference, payloadTypeName, propertyAndAnnotationCollector);
+        }
+
+        /// <summary>
+        /// Reads a resource value.
+        /// </summary>
+        /// <param name="structuredTypeReference">The expected type reference of the value.</param>
+        /// <param name="payloadTypeName">The type name read from the payload.</param>
+        /// <param name="propertyAndAnnotationCollector">The duplicate property names checker.</param>
+        /// <returns>The value of the resource value.</returns>
+        /// <remarks>
+        /// Pre-Condition:  JsonNodeType.Property - the first property of the resource value object, or the second one if the first one was odata.type.
+        ///                 JsonNodeType.EndObject - the end object of the resource value object.
+        /// Post-Condition: almost anything - the node after the resource value (after the EndObject)
+        /// </remarks>
+        private ODataResourceValue ReadResourceValue(

```

Comment:
>private ODataResourceValue ReadResourceValue [](start = 8, length = 44)

 This is the duplicated code that concerns me. What changes to regular resource would need to be made here as well?

Reply:
I had your concerns, but I think they are two different using scenarios. 
Basically, reading a value is very low level process, if we delegate the value reading to high level, in return, the high level will also call the low level process again. So, I am worry about the feasibility.

Of course, I did experiment using the top-level `ODataReader`. For example as below:

```C#
private ODataResourceValue ReadResourceValue(...)
{
   ODataReader reader = this.JsonLightInputContext.CreateUriParameterResourceReader(null, 
   structuredTypeReference == null ? null : structuredTypeReference.StructuredDefinition());
   return ReadResourceValue(reader);
 }

private static ODataResourceValue ReadResourceValue(ODataReader reader)
{
    while (reader.Read())
    {
         switch (reader.State)
          {
               case ODataReaderState.ResourceStart:
                   ....
          }
   }
}
```
but, So far, i can't make the above to work.

I think we are on the same page that we should have a method named "ReadResourceValue", the concern is about how to implement this private method.

Maybe we can postpone this detail implementation and focus on the step1 (adding class) and step2 (writing), hope we are on the same page about step1 and step2.

Reply:
If we have the right interface, we can optimize implementation later.

---
In reply to: [241615554](https://github.com/OData/odata.net/pull/1346#discussion_r241615554) [](ancestors = 241615554)